EIP-4399: Supplant DIFFICULTY opcode with RANDOM (#4399)
ethereum/EIPs: EIPS/eip-4399.md

Additionally, total difficulty computation would become overflow prone if the randomness output supplanted the value of the `difficulty` field.

### Reusing existing field instead of appending a new one

The `mixHash` field is deprecated at the PoS upgrade and set to zero bytes array thereafter. Reusing an existing field as a place for the randomness output saves 32 bytes per block and effectively removes the deprecation of one of the fields induced by the upgrade.

### Reusing the `DIFFICULTY` opcode instead of introducing a new one

See the [Motivation](#motivation).

### Renaming the field and the opcode

The renaming should be done to make the field and the opcode names semantically sound.

## Backwards Compatibility

This EIP doesn't introduce any backward incompatible changes to the base protocol. Though, the changes it propose might be backward incompatible for the following categories of applications:

> * Applications that uses the value returned by the `DIFFICULTY` opcode as the PoW `difficulty` parameter

Reviewer: ```suggestion
* Applications that use the value returned by the `DIFFICULTY` opcode as the PoW `difficulty` parameter
```

> * Applications which logic depending on the `DIFFICULTY` opcode return value to be relatively small number

Reviewer: ```suggestion
* Applications with logic that depends on the `DIFFICULTY` opcode returning a relatively small number
 ```

The first category is already affected by switching the consensus mechanism to PoS and no additional breaking changes are introduced by this specification.

> The second category comprises the applications that use the return value of the `DIFFICULTY` opcode in operations that might cause either overflow or underflow errors. Applications where the result of such errors may turn into security violations are not considered by this EIP.

Reviewer: ```suggestion
The second category is comprised of applications that use the return value of the `DIFFICULTY` opcode in operations that might cause either overflow or underflow errors.  While it is theoretically possible to author an application where a change in the range of possible values this OPCODE may return could lead to a security vulnerability, we believe the chances of that are very unlikely.
```

## Test Cases

Testing this functionality goes beyond this document and should be done in a form of integration testing.

## Reference Implementation

*TBD*

## Security Considerations

The `RANDOM` opcode in the PoS network should be considered as a source of randomness output of a higher strength than the output provided by either the `BLOCKHASH` or the `DIFFICULTY` opcode in the PoW network.

> However, the randomness output returned by the `RANDOM` opcode is revealed within the parent block.

Reviewer 2: yes

and an early randao reveal slashing is on the table
was in fact in the phase 0 specs early on!

but it's a relatively low priority attack point so I don't expect we see a proposal for inclusion in the next 18 months, tbh

> However, the randomness output returned by the `RANDOM` opcode is revealed within the parent block.

Author: My try to be more clear on it is: "is revealed to the network within the parent block" -- https://github.com/ethereum/EIPs/pull/4407. I slightly prefer us do not describe an attack scenario with the collusion of multiple proposers in this document. The reveal to the network should create enough level of awareness for smart contract developers. But I don't have a strong opinion on that and we could and it might worth specifying from some perspective.

Early randao reveal slashing doesn't protect revealing the randao mix, does it? If it does then revealing after grandparent block is publish can't be punished unless the proposer punishes itself.
